enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,249 @@
+- If the hinted SCC is not applicable, admission continues with asserting all the other SCCs available in the system,
+potentially even custom SCCs which still can cause problems.
+- All core OpenShift workloads need to set that annotation in their deployment artifacts.
+This implies that, after the change merges, many repositories have to be touched.
+- To verify that hinted SCCs are really being applied, an e2e test has to be written.
+This can be done centrally.
+
+An existing implementation is available at https://github.com/openshift/apiserver-library-go/pull/105.
+
+This proposal favors the "hard protection" variant and continues to describe that one further on.
+
+### Workflow Description
+
+Opting into this feature for core OpenShift workloads is straight-forward,
+with just an annotation needed to be set in the deployment manifest.
+
+### Variation (optional)
+
+Another variation for both proposed suggestions is to accept a comma-separated list of SCCs.
+This could be helpful in case additional SCCs are going to be introduced and the deployment manifest
+needs to be compatible with both old and new applicable SCCs.
+
+However, as adding a list further adds complexity, this suggestion is being omitted as an variation only
+with the potential of being introduced as a backwards-compatible change in the future.
+
+### API Extensions
+
+A new annotation `openshift.io/required-scc` is introduced.
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+### Risks and Mitigations
+
+1. OpenShift workloads that are not applicable to the requested SCC might fail to be admitted.
+Mitigation: The existing e2e test suite in OpenShift will give very early signals of failed admission attempts
+as the workload will be unschedulable.
+
+### Drawbacks
+
+This proposal adds yet another additional complexity on top of the existing SCC logic
+which also already interleaves with pod security admission.
+The mitigation here is to reduce complexity by favoring the "hard protection" variant.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. There might be benefit to limit the admission plugin to reconcile the new annotation in OpenShift namespaces only.
+Drawback: OpenShift workloads might be missed who wants to opt into this functionality.
+History shows that identifying "core OpenShift namespaces" is tricky.
+
+### Test Plan
+
+The existing OpenShift e2e test suite is assumed to be sufficient to verify correctness.
+If workloads opt into the annotation and force an SCC that is not applicable, admission will fail immediately.
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+The added annotation to OpenShift is ignored during downgrades.
+An upgrade strategy is not needed.
+
+When workloads are rescheduled during upgrades, admission will apply the newly added annotation.
```

Comment:
Should we take into account the possibility of a required SCC being accidentally deleted, which might not be immediately apparent due to other SCCs stepping in as substitutes?
This could lead to unintended consequences, as workloads might be admitted under different security constraints than expected.
It might be worth considering additional checks or validation steps to ensure that the required SCCs are present and active in the cluster, as well as mechanisms to restore default SCCs if needed.

Reply:
we already have this check in place: https://github.com/openshift/apiserver-library-go/blob/806f5687962951fefb4942064983bf1ed0cd9435/pkg/securitycontextconstraints/sccadmission/admission.go#L168

Reply:
SCC can be updated, and we will not reconcile them, but they will get recreated once they are removed.

Ftr: the above code exists mostly to delay SCC admission during cluster creation. Before it was done, workloads were failing to create with SCC errors during install because SCCs weren't there, which was an undesirable behavior.